include/onnxruntime/core/framework/sparse_tensor.h

```diff
@@ -1,72 +1,434 @@
+  /// <returns>nnz</returns>
   size_t NumValues() const { return static_cast<size_t>(values_.Shape().Size()); }
 
-  const Tensor& Values() const {
+  /// <summary>
+  /// Read only accessor to non-zero values
+  /// </summary>
+  /// <returns></returns>
+  const Tensor& Values() const noexcept {
     return values_;
   }
 
-  const Tensor& Indices() const {
-    return indices_;
+  SparseTensor(SparseTensor&& o) noexcept;
+  SparseTensor& operator=(SparseTensor&& o) noexcept;
+
+  /// <summary>
+  /// Returns SparseFormat that the instance currently holds
+  /// if the value returned in kUndefined, the instance is not populated
+  /// </summary>
+  /// <returns>format enum</returns>
+  SparseFormat Format() const noexcept {
+    return format_;
+  }
+
+  /// <summary>
+  /// Returns a would be dense_shape
+  /// </summary>
+  /// <returns>reference to dense_shape</returns>
+  const TensorShape& DenseShape() const noexcept {
+    return dense_shape_;
+  }
+
+  /// <summary>
+  /// Calculates and returns how much this fully initialized SparseTensor data (would)
+  /// occupy in a contiguous allocation block, or, in fact, occupies if it owns the buffer.
+  /// </summary>
+  /// <returns>required allocation size</returns>
+  int64_t RequiredAllocationSize() const noexcept;
+
+  /// <summary>
+  /// Returns Tensor element type enum.
+  /// Useful for type dispatching
+  /// </summary>
+  /// <returns></returns>
+  int32_t GetElementType() const {
+    return ml_data_type_->GetDataType();
   }
 
-  const TensorShape& Shape() const {
-    return shape_;
+  /// <summary>
+  /// Return Element MLDataType
+  /// </summary>
+  /// <returns></returns>
+  MLDataType DataType() const noexcept {
+    return ml_data_type_;
   }
 
-  Tensor& MutableValues() {
-    return values_;
+  /// <summary>
+  /// Test for string type
+  /// </summary>
+  /// <returns>true if tensor values are strings</returns>
+  bool IsDataTypeString() const {
+    return utils::IsPrimitiveDataType<std::string>(ml_data_type_);
   }
 
-  Tensor& MutableIndices() {
-    return indices_;
+  /// <summary>
+  /// Checks if the Tensor contains data type T
+  /// </summary>
+  /// <typeparam name="T"></typeparam>
+  /// <returns>true if tensor contains data of type T</returns>
+  template <class T>
+  bool IsDataType() const {
+    return utils::IsPrimitiveDataType<T>(ml_data_type_);
   }
 
-  //TensorShape& MutableShape() {
-  //  return shape_;
-  //}
+  const OrtMemoryInfo& Location() const noexcept { return location_; }
+
+  /// <summary>
+  /// Read only access to Coo indices
+  /// </summary>
+  class CooView {
+   public:
+    explicit CooView(const Tensor& indices) noexcept
+        : indices_(indices) {}
+    const Tensor& Indices() const noexcept { return indices_; }
+
+   private:
+    std::reference_wrapper<const Tensor> indices_;
+  };
+
+  /// <summary>
+  /// Returns Coo index view
+  /// </summary>
+  /// <returns>CooView instance</returns>
+  CooView AsCoo() const;
```

Comment:
May be it is obvious, but I guess these format-specific methods are valid only if the sparse-tensor has that format? How is the format set? May help to document.

Reply:
The format is set when Use*() or Make*() methods for a specific format initialize format specific indices.